google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -43,90 +44,117 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
     message_span->SetAttribute("pubsub.batch_sink.span_id", span_id);
   }
+  return batch_sink_parent;
+}
 
-  return batch_sink_parent_span;
+auto MakeChild(
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> const& parent,
+    int count, Links const& links) {
+  opentelemetry::trace::StartSpanOptions options;
+  options.parent = parent->GetContext();
+  return internal::MakeSpan(
+      "BatchSink::AsyncPublish - Batch #" + std::to_string(count),
+      /*attributes=*/{{}},
+      /*links=*/links, options);
+}
+
+Spans MakeBatchSinkSpans(Spans message_spans) {
+  int constexpr kMaxOtelLinks = 128;
+  Spans batch_sink_spans;
+  // If the batch size is less than the max size, add the links to a single
+  // span. If the batch size is greater than the max size, create a parent
+  // span with no links and each child spans will contain links.
+  if (message_spans.size() <= kMaxOtelLinks) {
+    batch_sink_spans.push_back(MakeParent(
+        MakeLinks(message_spans.begin(), message_spans.end()), message_spans));
+    return batch_sink_spans;
+  }
+  batch_sink_spans.push_back(MakeParent({{}}, message_spans));
+  auto batch_sink_parent = batch_sink_spans.front();
+
+  auto cut = [&message_spans](auto i) {
+    auto const batch_size = static_cast<std::ptrdiff_t>(kMaxOtelLinks);
+    return std::next(
+        i, std::min(batch_size, std::distance(i, message_spans.end())));
+  };
+  int count = 0;
+  for (auto i = message_spans.begin(); i != message_spans.end(); i = cut(i)) {
+    // Generates child spans with links between [i, min(i + batch_size, end))
+    // such that each child span will have exactly batch_size elements or less.
+    batch_sink_spans.push_back(
+        MakeChild(batch_sink_parent, count++, MakeLinks(i, cut(i))));
+  }
+
+  return batch_sink_spans;
 }
 
 }  // namespace
 
-void TracingMessageBatch::Flush() {
+future<void> TracingMessageBatch::Flush() {
   decltype(message_spans_) message_spans;
   {
     std::lock_guard<std::mutex> lk(message_mu_);
     message_spans.swap(message_spans_);
   }
 
-  // TODO(#12528): Handle batches larger than 128.
-  auto batch_sink_parent_span = MakeParentSpan(std::move(message_spans));
+  auto batch_sink_spans = MakeBatchSinkSpans(std::move(message_spans));
 
-  // Set the batch sink as the active span.
+  // Set the batch sink parent span as the active span. This will always be the
+  // first span in the vector.
   auto async_scope = internal::GetTracer(internal::CurrentOptions())
-                         ->WithActiveSpan(batch_sink_parent_span);
-  {
-    std::lock_guard<std::mutex> lk(batch_sink_mu_);
-    batch_sink_spans_.push_back(std::move(batch_sink_parent_span));
-  }
+                         ->WithActiveSpan(batch_sink_spans.front());
 
-  child_->Flush();
+  return child_->Flush().then([spans = std::move(batch_sink_spans)](auto f) {
+    for (auto& span : spans) {
+      internal::EndSpan(*span);
+    }
+  });
```

Comment:
Maybe you no longer need `FlushCallback()`.  In `BatchingPublisherConnection::FlushImpl()` you have code like this:

```cc
  batch_->Flush();
  batch.batch = batch_;
  sink_->AsyncPublish(std::move(request)).then(std::move(batch));
```

I think that can be changed to something more like this:

```cc
  std::function<void(future<void>)> handler = batch_->Flush();
  sink_->AsyncPublish(std::move(request)).then(std::move(batch)).then(std::move(handler));
```

which would require changing the code here to:


```suggestion
  return [next = child_->Flush(), spans = std::move(batch_sink_spans)](auto) {
    for (auto& span : spans) internal::EndSpan(*span);
    next();
  };
```

Reply:
I agree. And I think this is a lot cleaner.

Thanks Carlos!